enable support for nested containers in `torch.testing.assert(equal|close)`

pmeier opened the pull request:
In contrast to the initial opinion in #55385, there are legitimate use cases for nested containers. One such example is the [output of `LSTM`'s](https://pytorch.org/docs/stable/generated/torch.nn.LSTM):

```python
output: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]] = torch.nn.LSTM()(input)
assert_close(output, expected)
```

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-            f"Additional keys in the actual mapping: {sorted(additional_keys)}\n"
-        )
-        return exc, None
+    # We explicitly exclude str's here since they are self-referential and would cause an infinite recursion loop:
+    # "a" == "a"[0][0]...
```
Of course the same is true for other self-referential sequences. Safeguarding this for the general case is more involved and should only be implemented if there is an actual use case.

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
 
-def _parse_mappings(actual: Mapping, expected: Mapping) -> Tuple[Optional[Exception], Optional[_ParsedInputs]]:
-    """Parses sequences of scalar-or-array-like pairs.
+def _parse_inputs(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[Union[_TensorPair, List, Dict]]]:
```
Before we returned two corresponding tensors or tensor containers. Now this function bundles the actual and expected tensors together into one `_TensorPair`. I plan to use this for all internal functions that happen "after" the input parsing.

mruberry replied:
What's the benefit of pairing them in an object vs. just passing actual and expected directly?

pmeier replied:
Especially when having nested containers, it is harder two maintain two inputs with 1-to-1 correspondence instead of one where each item is a pair of two tensors. This function performs this transformation.

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
+            exc = AssertionError(f"The length of the sequences mismatch: {actual_len} != {expected_len}")
+            return exc, None
+
+        pair_list = []
+        for idx in range(actual_len):
+            exc, pair = _parse_inputs(actual[idx], expected[idx])
+            if exc:
+                exc = _amend_error_message(exc, f"{{}}\n\n{_SEQUENCE_MSG_FMTSTR.format(idx)}")
+                return exc, None
```
I've inlined this from `_parse_sequences`. Although this makes this function harder to read, I wanted to avoid alternating recursive call between two functions.

pmeier commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-        actual_dct[key] = cast(Tensor, result.actual)
-        expected_dct[key] = cast(Tensor, result.expected)
+        pair_dict = {}
+        for key in sorted(actual_keys):
+            exc, pair = _parse_inputs(actual[key], expected[key])
+            if exc:
+                exc = _amend_error_message(exc, f"{{}}\n\n{_MAPPING_MSG_FMTSTR.format(key)}")
+                return exc, None
```
Same as above.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-    if isinstance(actual, collections.abc.Sequence) and isinstance(expected, collections.abc.Sequence):
-        for idx, (actual_t, expected_t) in enumerate(zip(actual, expected)):
-            exc = check_tensors(actual_t, expected_t)
+    if isinstance(pair, list):
```
Why the change from sequence to list? What about tuples?

pmeier replied:
We could test for sequences and mappings here, but `_parse_inputs` transforms each sequence to a list and each mapping to a dict.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -370,194 +370,181 @@ def _amend_error_message(exc: E, msg_fmtstr: str) -> E:
-    """Parses sequences of scalar-or-array-like pairs.
 
-    Regardless of the input types, the sequences are returned as :class:`list`.
+def _to_tensor_pair(actual: Any, expected: Any) -> Tuple[Optional[Exception], Optional[_TensorPair]]:
```
It does seem like there's a lot of architecture churn to support this scenario, we probably want to review the architecture holistically before 1.9, if possible

pmeier replied:
We already had this before: `_parse_array_or_scalar_like_pair` on the old L443. We always need to parse a pair to give expressive messages for offending index / key in case of containers.

mruberry reviewed (commented):
Hey @pmeier, overall this seems cool, although I'd like to be sure we're continuing to support tuples, too, for example. I made a few inline comments. Let's give @mattip a chance to take a look, too.

Looking forward to hearing your thoughts!

mattip commented:
I would prefer not to do this until end-users request it. In the case of LSTM, it can be avoided by using the construct as per the example in the documentation, which has the advantage of clearly indicating which return value is not correct, at the cost of erroring out on the first failure:
```
>>> output, (hn, cn) = rnn(input, (h0, c0))
>>> assert_close(output, expected_out)
>>> assert_close(hn, expected_hn)
>>> assert_close(cn, expected_cn)
```

Is there a more compelling reason to add this?

pmeier commented:
@mattip In general I agree with you. But in some places in the test suite this makes the tests less clear than before:

https://github.com/pytorch/pytorch/blob/7e12c3e10a64530d34d760f1b9bdeb84e32979a9/test/quantization/test_backward_compatibility.py#L219-L230

Here `self._test_op()` is a test helper that is also called by many other tests. Thus we would have two options here:

1. Make `LSTM` a special case and re-implement the functionality of `self._test_op()` for it.
2. Adapt `self._test_op()` to perform container unpacking to be able to handle all cases.

Both make the test more complex and harder to maintain.

---

@mruberry  and I discussed to maybe add a warning to error message in case we detect nested containers:

```
AssertionError: [...] The error message is hard to parse, because you are testing nested containers. To improve this we suggest refactoring your test.
```

What do you think?

mattip commented:
I personally was against adding containers to `testing.assert_equal`, so my position for nested containers is even more strongly negative. I think a single test does not justify the added complications stemming from this PR. The test should be rewritten to avoid returning a nested container. I think the test is supposed to be checking serialization, it  can check that on a different op that does not need nested containers. If the goal of the test is to check serialization of nested containers, then the test should be made more complicated, not `testing.assert_equal`.

pmeier commented:
> I personally was against adding containers to `testing.assert_equal`

There are clear-cut uses cases for tensor containers, for example checking the state dicts of `nn.Module`'s. Forcing the user to unpack it IMO makes the test less clear. Plus, for this scenario the error message is very expressive since we append to offending index / key to it.

> The test should be rewritten to avoid returning checking a nested container.

I agree, but I'm not sure if its worth the effort. I only gave one example above. There are quite a few more to fix if we really want to do that. See #56472 for an "overview".

mattip reviewed (commented):
.

mruberry reviewed (approved):
Cool! This will need a rebase; maybe it should wait for the testing and docs PRs to land?

pmeier committed:
enable support for nested containers in `torch.testing.assert(equal|close)`

In contrast to the initial opinion in #55385, there are legitimate use cases for
nested containers. One such example is the [output of
`LSTM`'s](https://pytorch.org/docs/stable/generated/torch.nn.LSTM):

```python
output: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]] =
torch.nn.LSTM()(input)
assert_close(output, expected)
```
